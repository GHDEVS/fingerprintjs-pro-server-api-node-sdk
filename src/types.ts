export enum Region {
  EU = 'EU',
  AP = 'AP',
  Global = 'Global',
}

export enum AuthenticationMode {
  AuthHeader = 'AuthHeader',
  QueryParameter = 'QueryParameter',
}

/**
 * Options for FingerprintJS server API client
 */
export interface Options {
  /**
   * Secret API key
   */
  apiKey: string;
  /**
   * Region of the FingerprintJS service server
   */
  region: Region;
  /**
   * Authentication mode
   * Optional, default value is AuthHeader
   */
  authenticationMode?: AuthenticationMode;
}

export type VisitorHistoryFilter = {
  request_id?: string;
  linked_id?: string;
  limit?: number;
  before?: number;
};

export interface VisitorsResponse {
  lastTimestamp?: string;
  visitorId: string;
  visits: Visit[];
}

export interface Visit {
  requestId: string;
  incognito: string;
  linkedId: 'string';
  time: Date;
  timestamp: number;
  url: string;
  ip: string;
  ipLocation: IpLocation;
  browserDetails: BrowserDetails;
}

export interface IpLocation {
  accuracyRadius: number;
  latitude: number;
  longitude: number;
  postalCode: string;
  timezone: string;
  city: City;
  continent: Continent;
  country: Country;
  subdivisions: Subdivision[];
}

export interface City {
  name: string;
}

export interface Continent {
  code: string;
  name: string;
}

export interface Country {
  code: string;
  name: string;
}

export interface Subdivision {
  isoCode: string;
  name: string;
}

export interface BrowserDetails {
  browserName: string;
  browserMajorVersion: string;
  browserFullVersion: string;
  os: string;
  osVersion: string;
  device: string;
  userAgents: string;
}

export interface VisitWebhook {
  // Unique request identifier
  // nullable: false, maxLength: 20
  requestId: string;
  // customer-provided data, for example requestType and yourCustomId
  // both the tag and the information it contains are optional
  // and only for the customer's need.
  // nullable: true, maxLength: 4096
  tag: any | null;
  // user-provided scalar identifier
  // nullable: true, maxLength: 4096
  linkedId: string | null;
  // persistent visitor identifier
  // helpful to detect anonymous or private mode users
  // nullable: false, maxLength: 20
  visitorId: string;
  // timestamp in milliseconds (since unix epoch)
  // nullable: false
  timestamp: number;
  // if the page view was made in incognito or private mode
  // nullable: false
  incognito: boolean;
  // URL where the API was called in the browser
  // nullable: false, maxLength: 4096
  url: string;
  // The URL of a client-side referrer
  // nullable: true, maxLength: 4096
  clientReferrer: string | null;
  // nullable: false, maxLength: 40
  ip: string;
  // nullable: false
  ipLocation: {
    // miles
    // nullable: true
    accuracyRadius: number | null;
    // nullable: true
    city: {
      // nullable: true, maxLength: 4096
      name: string | null;
    } | null;
    // nullable: true
    continent: {
      // nullable: true, maxLength: 2
      code: string | null;
      // nullable: true, maxLength: 40
      name: string | null;
    } | null;
    // nullable: true
    country: {
      // nullable: true, maxLength: 2
      code: string | null;
      // nullable: true, maxLength: 250
      name: string | null;
    } | null;
    // nullable: true
    latitude: number | null;
    // nullable: true
    longitude: number | null;
    // nullable: true, maxLength: 40
    postalCode: string | null;
    // nullable: true
    subdivisions:
      | {
          // nullable: true, maxLength: 250
          isoCode: string | null;
          // nullable: true, maxLength: 250
          name: string | null;
        }[]
      | null;
    // nullable: true, maxLength: 250
    timezone: string | null;
  };
  // nullable: false
  browserDetails: {
    // nullable: true, maxLength: 250
    browserName: string | null;
    // nullable: true, maxLength: 250
    browserFullVersion: string | null;
    // nullable: true, maxLength: 250
    browserMajorVersion: string | null;
    // nullable: true, maxLength: 250
    os: string | null;
    // nullable: true, maxLength: 250
    osVersion: string | null;
    // nullable: true, maxLength: 250
    device: string | null;
    // nullable: true, maxLength: 4096
    userAgent: string | null;
  };
}
