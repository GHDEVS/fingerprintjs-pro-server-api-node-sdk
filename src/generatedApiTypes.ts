/**
 * This file was auto-generated by openapi-typescript.
 * Do not make direct changes to the file.
 */

export interface paths {
  '/visitors/{visitor_id}': {
    /** This endpoint allows you to get a history of visits with all available information. Use the visitorID as a URL path parameter. This API method is scoped to a visitor, i.e. all returned information is by visitorId. */
    get: {
      parameters: {
        path: {
          visitor_id: string;
        };
        query: {
          /** Filter events by requestId */
          request_id?: string;
          /** Filter events by custom identifier */
          linked_id?: string;
          /** Limit scanned results */
          limit?: number;
          /** Used to paginate results */
          before?: number;
        };
      };
      responses: {
        /** Auto generated using Swagger Inspector */
        200: {
          content: {
            'application/json': components['schemas']['Response'];
          };
        };
        429: {
          content: {
            'application/json': components['schemas']['ManyRequestsResponse'];
          };
        };
      };
    };
  };
  '/webhook': {
    /** Fake path to describe webhook format. More information about webhooks can be found in the [documentation](https://dev.fingerprint.com/docs/webhooks) */
    trace: {};
  };
}

export interface components {
  schemas: {
    Response: components['schemas']['BaseResponse'] | components['schemas']['PaginatedResponse'];
    /** BaseResponse */
    BaseResponse: {
      visitorId: string;
      visits: components['schemas']['Visit'][];
    };
    /**
     * PaginatedResponse
     * @description Fields `lastTimestamp` and `paginationKey` added when `limit` or `before` parameter provided and there is more data to show
     */
    PaginatedResponse: {
      visitorId: string;
      visits: components['schemas']['Visit'][];
      /**
       * Format: int64
       * @example 1654815517198
       */
      lastTimestamp: number;
      /** @example 1654815517198.azN4IZ */
      paginationKey: string;
    };
    ManyRequestsResponse: {
      /** @example request throttled */
      error: string;
    };
    WebhookVisit: {
      visitorId: string;
      clientReferrer?: string;
    } & components['schemas']['Visit'];
    /** Visit */
    Visit: {
      /** @example 1654815516083.OX6kx8 */
      requestId: string;
      browserDetails: components['schemas']['BrowserDetails'];
      incognito: boolean;
      /**
       * Format: ipv4
       * @example 8.8.8.8
       */
      ip: string;
      ipLocation: components['schemas']['IPLocation'];
      /**
       * Format: int64
       * @example 1654815516086
       */
      timestamp: number;
      /**
       * Format: date-time
       * @example 2022-06-09T22:58:36Z
       */
      time: string;
      /**
       * Format: uri
       * @example https://some.website/path?query=params
       */
      url: string;
      tag: { [key: string]: unknown };
      /** @example someID */
      linkedId?: string;
      confidence: components['schemas']['Confidence'];
      visitorFound: boolean;
      /**
       * @example {
       *   "global": "2022-05-05T18:28:54.535Z",
       *   "subscription": "2022-06-09T22:58:05.576Z"
       * }
       */
      firstSeenAt: components['schemas']['StSeenAt'];
      /**
       * @example {
       *   "global": "2022-06-09T22:58:35.795Z",
       *   "subscription": null
       * }
       */
      lastSeenAt: components['schemas']['StSeenAt'];
    };
    /** BrowserDetails */
    BrowserDetails: {
      /** @example Chrome */
      browserName: string;
      /** @example 101 */
      browserMajorVersion: string;
      /** @example 101.0.4951 */
      browserFullVersion: string;
      /** @example Windows */
      os: string;
      /** @example 10 */
      osVersion: string;
      /** @example Other */
      device: string;
      /** @example Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.41 Safari/537.36 */
      userAgent: string;
      botProbability?: number;
    };
    /** Confidence */
    Confidence: {
      /** Format: float */
      score: number;
    };
    /** StSeenAt */
    StSeenAt: {
      /** Format: date-time */
      global: string;
      /** Format: date-time */
      subscription: string | null;
    };
    /** IPLocation */
    IPLocation: {
      /** @example 1000 */
      accuracyRadius: number;
      /**
       * Format: double
       * @example 37.75
       */
      latitude: number;
      /**
       * Format: double
       * @example -97.82
       */
      longitude: number;
      /** @example 130 00 */
      postalCode?: string;
      /**
       * Format: timezone
       * @example America/Chicago
       */
      timezone: string;
      city?: {
        /** @example Prague */
        name?: string;
      };
      /**
       * @example {
       *   "code": "US",
       *   "name": "United States"
       * }
       */
      country: components['schemas']['Location'];
      /**
       * @example {
       *   "code": "NA",
       *   "name": "North America"
       * }
       */
      continent: components['schemas']['Location'];
      subdivisions?: components['schemas']['Subdivision'][];
    };
    /** Continent */
    Location: {
      /** @example US */
      code: string;
      /** @example United States */
      name: string;
    };
    Subdivision: {
      /** @example 10 */
      isoCode?: string;
      /** @example Hlavni mesto Praha */
      name?: string;
    };
  };
}

export interface operations {}

export interface external {}
