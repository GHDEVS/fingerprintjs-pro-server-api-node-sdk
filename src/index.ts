export * from  "./urlUtils";
export * from  "./serverApiClient";
export * from  "./serverApiConfig";
export * from  "./types";