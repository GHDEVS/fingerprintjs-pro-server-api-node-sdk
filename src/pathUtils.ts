import { Region } from "./types";

const euRegionUrl = "https://eu.api.fpjs.io/";
const globaRegionUrl = "https://api.fpjs.io/";

export function getVisitorsPath(region: Region, visitorId: string): string {
    const serverApiUrl = getServerApiUrl(region);
    const serverApiPath = `${serverApiUrl}visitors/${visitorId}`;
    return serverApiPath;
  }

  export function getServerApiUrl(region: Region): string {
    switch (region) {
      case Region.EU:
        return euRegionUrl;
      case Region.Global:
        return globaRegionUrl;
      default:
        throw new Error(`Unsupported region`);
    }
  }