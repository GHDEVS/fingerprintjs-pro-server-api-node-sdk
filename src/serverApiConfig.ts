import { Region } from "./types";

export class FingerprintJsServerApiConfig {
    public readonly region: Region;
    public readonly authToken: string;
  
    constructor(region: Region, apiToken: string) {
      if (!region) {
        throw Error(`Region is not set`);
      }
  
      if (!apiToken) {
        throw Error(`Api token is not set`);
      }
  
      this.region = region;
      this.authToken = apiToken;
    }
  }
  