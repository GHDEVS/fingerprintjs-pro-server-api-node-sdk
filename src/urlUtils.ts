import { Region, VisitorHistoryFilter } from "./types";
import querystring from 'querystring'

const euRegionUrl = "https://eu.api.fpjs.io/";
const globaRegionUrl = "https://api.fpjs.io/";

export function getVisitorsUrl(region: Region, visitorId: string, filter?: VisitorHistoryFilter): string {
  const serverApiPath = getVisitorsPath(region, visitorId);
  const queryString = filter ? querystring.stringify(filter) : '';

  if (queryString === '') {
    return serverApiPath;
  } else {
    return `${serverApiPath}?${queryString}`;
  }
}

function getVisitorsPath(region: Region, visitorId: string): string {
  const serverApiUrl = getServerApiUrl(region);
  const serverApiPath = `${serverApiUrl}visitors/${visitorId}`;
  return serverApiPath;
}

function getServerApiUrl(region: Region): string {
  switch (region) {
    case Region.EU:
      return euRegionUrl;
    case Region.Global:
      return globaRegionUrl;
    default:
      throw new Error(`Unsupported region`);
  }
}