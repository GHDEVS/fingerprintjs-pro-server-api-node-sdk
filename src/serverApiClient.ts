import fetch from 'node-fetch';
import { getVisitorsUrl } from './urlUtils';
import { FingerprintJsServerApiConfig } from './serverApiConfig';
import { VisitorHistoryFilter, VisitorsResponse } from './types';

export class FingerprintJsServerApiClient {
  private readonly clientConfig: FingerprintJsServerApiConfig;

  /**
  * FingerprintJS server API client used to fetch data from FingerprintJS
  * @constructor
  * @param {FingerprintJsServerApiConfig} config - The client configuration
  */
  constructor(config: FingerprintJsServerApiConfig) {
    if (!config) {
      throw new Error(`Config is not set`);
    }

    this.clientConfig = config;
  }

  /**
  * Gets history for the given visitor
  * @param {string} visitorId - Identifier of the visitor
  * @param {VisitorHistoryFilter} filter - Visitor history filter
  */
  public async getVisitorHistory(visitorId: string, filter?: VisitorHistoryFilter): Promise<VisitorsResponse> {
    if (!visitorId) {
      throw Error(`VisitorId is not set`);
    }

    const url = getVisitorsUrl(this.clientConfig.region, visitorId, filter);

    return fetch(url, {
      method: 'GET',
      headers: { 'Auth-Token': this.clientConfig.apiToken },
    })
      .then((response) => {
        return response.json()
          .then(jsonData => {
            const visitorsResponse = jsonData as VisitorsResponse;
            return visitorsResponse;
          })
          .catch(err => {
            throw new Error(err.toString());
          })
      })
      .catch((err) => {
        throw new Error(err.toString());
      })
  }
}