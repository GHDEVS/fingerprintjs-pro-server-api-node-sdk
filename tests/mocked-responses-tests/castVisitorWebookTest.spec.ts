import { VisitWebhook } from '../../src/types';
import visitWebhookBody from './mocked-responses-data/visit-webhook-body.json';

describe('[Mocked body] Cast visitor webhook', () => {
  test('with sample request body', async () => {
    const visit = visitWebhookBody as unknown as VisitWebhook;

    expect(visit).toMatchSnapshot();

    // TODO: check all properties manually
    // expect(requestBody.requestId).toEqual("1627054081281.wo4lLc");
    // expect(requestBody.visitorId).toEqual("TaDnMBz9XCpZNuSzFUqP");

    // expect(requestBody.browserDetails.browserName).toEqual("Chrome");
    // expect(requestBody.browserDetails.browserMajorVersion).toEqual("92");
    // expect(requestBody.browserDetails.browserFullVersion).toEqual("92.0.4515");
    // expect(requestBody.browserDetails.os).toEqual("Windows");
    // expect(requestBody.browserDetails.osVersion).toEqual("10");
    // expect(requestBody.browserDetails.device).toEqual("Other");
    // expect(requestBody.browserDetails.userAgent).toEqual("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.107 Safari/537.36");
    // // not documented
    // // expect(requestBody.browserDetails.botProbability).toEqual(0);

    // expect(requestBody.incognito).toEqual(false);
    // expect(requestBody.ip).toEqual("185.5.70.59");

    // expect(requestBody.ipLocation.accuracyRadius).toEqual(100);
    // expect(requestBody.ipLocation.city).toEqual(null);
  });
});
