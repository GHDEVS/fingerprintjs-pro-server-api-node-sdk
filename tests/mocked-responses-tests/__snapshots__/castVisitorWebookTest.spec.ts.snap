// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Mocked body] Cast visitor webhook with sample request body 1`] = `
Object {
  "browserDetails": Object {
    "botProbability": 0,
    "browserFullVersion": "92.0.4515",
    "browserMajorVersion": "92",
    "browserName": "Chrome",
    "device": "Other",
    "os": "Windows",
    "osVersion": "10",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.107 Safari/537.36",
  },
  "clientReferrer": "https: //www.martinmakarsky.com/",
  "incognito": false,
  "ip": "185.5.70.59",
  "ipLocation": Object {
    "accuracyRadius": 100,
    "city": Object {},
    "continent": Object {
      "code": "EU",
      "name": "Europe",
    },
    "country": Object {
      "code": "CZ",
      "name": "Czechia",
    },
    "latitude": 50.0848,
    "longitude": 14.4112,
    "organization": Object {},
    "timezone": "Europe/Prague",
  },
  "linkedId": "makma",
  "requestId": "1627054081281.wo4lLc",
  "tag": Object {
    "tag": "my_tag",
  },
  "time": "",
  "timestamp": 1627054081282,
  "url": "https://www.martinmakarsky.com/not-very-secret-experiment-page",
  "visitorId": "TaDnMBz9XCpZNuSzFUqP",
}
`;
